### PR TITLE
make const_err a hard error

### LINKED ISSUE #71800: Make const_err a future-incompatibility lint and eventually a hard error
Currently, when Miri raises an error while evaluating a `const`, that does *not* lead to a hard compiler error (`sess.struct_err` or so). Instead, the `const_err` lint is emitted (and there are a bunch of exceptions I keep forgetting... promoteds and `static` each get their own treatment). Then later consumers of consts have to remember to `struct_err` when they query for the value of the const (and they do that with inconsistent messages and some might even forget to do it), usually without any actual details so if you `allow(const_err)` the error makes no sense... in a word, it's a mess (and so is the code backing this).

Can we clean that up? I think we should try. IMO the long-term goal should be to just make errors when evaluating a const hard errors, so we can have that error-reporting code once in `librustc_mir/const_eval` and not worry about it everywhere else in the compiler. Right now, it's a deny-by-default lint, for backwards compatibility.

To make this slightly more complicated, we also [evaluate consts that the program does not actually use](https://github.com/rust-lang/rust/blob/master/src/librustc_lint/builtin.rs#L1161-L1188), so we can tell library crate authors when their consts are wrong. Inside the const-eval engine, these look like normal const-eval queries, so if we switch to a hard error, that would be a hard error, too. We have to decide what to do about that.

### Option 1: treat unused consts like used consts whenever we can

The easiest way forward would be to say that an eventual hard error for CTFE failures in unused consts is fine. Then we can just make the current `const_err` lint reporting into a future-incompatibility lint (maybe clean it up a little, and clean up the way that we error when we truly need the const value). Later, we can experiment with making it a hard error, and at that point everything would become simple and uniform and beautiful. (Promoteds would still only lint as the user did not ask for const-evaluation, but at least behavior wouldn't depend on context any more: const/static *always* hard-error [which currently const only does for some callers], and promoteds *never* do. Okay maybe it's not beautiful but it is better.)

Note that this matches what we already do for `static`. However, unlike `static`, `const` can exist in generic contexts (as associated consts), and those we cannot evaluate and thus not lint or error for. So, "top-level" `const` would behave differently from associated consts in terms of error reporting. (That is already the case, but currently the difference is a deny-by-default lint that can be allow'ed away.) There is no technical *need* to hard-error on CTFE failures in unused consts, but then one could say the same about unused statics and we *do* already hard-error there (and I don't remember anyone ever complaining about that).  So my personal preference would be to hard-error for failing consts as often as we can.

### Option 2: keep just linting for unused consts

Alternatively, maybe we don't want to hard-error when a const that is not used fails to evaluate (consistently with how we *cannot* produce a hard error when an assoc const fails to evaluate as we cannot even check that). Arguably we should then downgrade such messages to warn-by-default.

In this case we should probably find a way for the caller to inform the const-eval engine whether they truly need the result of the query or not (almost always they do need it, except for that lint), and then use that to control our lint level... but how can we do that without duplicating errors or work? Currently it is deduplicated because the query only runs once, and it always emits a lint, and then the caller potentially additionally emits a hard error but pointing at the span where the const actually gets used. But this is bad, it means that with `allow(const_err)` none of the relevant error details are shown even when we do get a hard error.

### Backwards compatibility concerns

A potential concern here might be that if we make CTFE errors hard errors, we have to be careful with making the Miri engine raise more CTFE errors (like, improved UB checking). But that is *already* the case: when raise more CTFE errors, the consts affected by it go from having a value to not having one, and cannot be used any more. So I don't think the proposal here creates any new backcompat issues.

Cc @rust-lang/wg-const-eval @rust-lang/lang

### LINKED ISSUE #100114: `no warnings or errors encountered even though delayed_good_path_bugs issued` when constructing a `const` never type with warnings disabled
I was working on some perfectly reasonable and definitely not cursed Rust code when Cargo suddenly started crashing for some reason.

### Code

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=18d868f2895c570c6e7d167b73d0fe42)

```Rust
#![allow(warnings)]
#![feature(never_type)]
#![allow(const_err)]

use std::mem::MaybeUninit;

const fn never() -> ! {
    unsafe { MaybeUninit::uninit().assume_init() }
}

const NEVER: ! = never();

fn main() {}

```

### Meta
This happens on nightly

`cargo +nightly --version`:
```
cargo 1.64.0-nightly (85b500cca 2022-07-24)
```

### Error output

```
error: internal compiler error: no warnings or errors encountered even though `delayed_good_path_bugs` issued
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: no warnings or errors encountered even though `delayed_good_path_bugs` issued

error: internal compiler error: trimmed_def_paths constructed
  |
  = note: delayed at    0: <rustc_errors::Handler>::delay_good_path_bug::<&str>
             1: rustc_middle::ty::print::pretty::trimmed_def_paths
             2: <rustc_middle::dep_graph::dep_node::DepKind as rustc_query_system::dep_graph::DepKind>::with_deps::<<rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task_impl<rustc_middle::ty::context::TyCtxt, (), std::collections::hash::map::HashMap<rustc_span::def_id::DefId, rustc_span::symbol::Symbol, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>::{closure#0}, std::collections::hash::map::HashMap<rustc_span::def_id::DefId, rustc_span::symbol::Symbol, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>
             3: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), std::collections::hash::map::HashMap<rustc_span::def_id::DefId, rustc_span::symbol::Symbol, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>
             4: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::ArenaCache<(), std::collections::hash::map::HashMap<rustc_span::def_id::DefId, rustc_span::symbol::Symbol, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>>
             5: rustc_query_system::query::plumbing::force_query::<rustc_query_impl::queries::trimmed_def_paths, rustc_query_impl::plumbing::QueryCtxt>
             6: rustc_query_impl::query_callbacks::trimmed_def_paths::force_from_dep_node
             7: <rustc_middle::ty::context::TyCtxt as rustc_query_system::dep_graph::DepContext>::try_force_from_dep_node
             8: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl::plumbing::QueryCtxt>
             9: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl::plumbing::QueryCtxt>
            10: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl::plumbing::QueryCtxt>
            11: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::try_mark_previous_green::<rustc_query_impl::plumbing::QueryCtxt>
            12: rustc_query_system::query::plumbing::ensure_must_run::<rustc_query_impl::plumbing::QueryCtxt, rustc_span::def_id::LocalDefId, ()>
            13: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::lint_mod, rustc_query_impl::plumbing::QueryCtxt>
            14: <rustc_middle::hir::map::Map>::for_each_module::<rustc_lint::late::check_crate<rustc_lint::BuiltinCombinedLateLintPass, rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}::{closure#0}>::{closure#1}::{closure#0}::{closure#0}>
            15: <rustc_session::session::Session>::time::<(), rustc_lint::late::check_crate<rustc_lint::BuiltinCombinedLateLintPass, rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}::{closure#0}>::{closure#1}::{closure#0}>
            16: <rustc_session::session::Session>::time::<(), rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}>
            17: <core::panic::unwind_safe::AssertUnwindSafe<rustc_interface::passes::analysis::{closure#5}::{closure#1}> as core::ops::function::FnOnce<()>>::call_once
            18: <rustc_session::session::Session>::time::<(), rustc_interface::passes::analysis::{closure#5}>
            19: rustc_interface::passes::analysis
            20: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            21: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorGuaranteed>>>
            22: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
            23: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorGuaranteed>>
            24: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
            25: rustc_interface::interface::create_compiler_and_run::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>
            26: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            27: std::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            28: <<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
            29: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                       at /rustc/4493a0f4724c0bae1436242d76cccc9c0a287b80/library/alloc/src/boxed.rs:1935:9
            30: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                       at /rustc/4493a0f4724c0bae1436242d76cccc9c0a287b80/library/alloc/src/boxed.rs:1935:9
            31: std::sys::unix::thread::Thread::new::thread_start
                       at /rustc/4493a0f4724c0bae1436242d76cccc9c0a287b80/library/std/src/sys/unix/thread.rs:108:17
            32: start_thread
                       at /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477:8
            33: clone
                       at /build/glibc-SzIz7B/glibc-2.31/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
          

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1426:13
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: std::panic::panic_any::<rustc_errors::ExplicitBug>
   2: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
   3: core::ptr::drop_in_place::<rustc_session::parse::ParseSess>
   4: <alloc::rc::Rc<rustc_session::session::Session> as core::ops::drop::Drop>::drop
   5: core::ptr::drop_in_place::<rustc_interface::interface::Compiler>
   6: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
   7: rustc_interface::interface::create_compiler_and_run::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>
   8: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.64.0-nightly (4493a0f47 2022-08-02) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C embed-bitcode=no -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `rust_playground`; 1 warning emitted
```

</p>
</details>

### PR DESCRIPTION
**Pull request by RalfJung**
This lint has been deny-by-default with future incompat wording since [Rust 1.51](https://github.com/rust-lang/rust/pull/80394) and the stable release of this week starts showing it in cargo's future compat reports. I can't wait to finally get rid of at least some of the mess in our const-err-reporting-code. ;)

r? @oli-obk 
Fixes https://github.com/rust-lang/rust/issues/71800
Fixes https://github.com/rust-lang/rust/issues/100114

### CODE REVIEW

File: src/test/ui/lint/force-warn/allowed-cli-deny-by-default-lint.rs [hunk 1]
```diff
@@ -1,10 +1,9 @@
...
 
-const C: i32 = 1 / 0;
-//~^ WARN any use of this value will cause an error
-//~| WARN this was previously accepted by the compiler
-
-fn main() {}
+fn main() {
+    let _val = 1_i32 << 32;
+    //~^WARN: attempt to shift left by `32_i32`, which would overflow
+}
```
Comment (RalfJung) [inline src/test/ui/lint/force-warn/allowed-cli-deny-by-default-lint.rs:9]: @rylev @ehuss (who edited this file in the past) I don't have a clue what this is testing, but it has to switch from const_err to something else. (It's not a good idea to use a lint for such a test that is planned to be a hard error in the future. Just introduces more work for whoever is doing that transition eventually.)  From the comments it sounds like it needs to be a deny-by-default lint. `arithmetic_overflow` is a deny-by-default lint but here it does not behave the way the test wants it to. No warning is printed. I am at a loss.

Comment (RalfJung) [reply src/test/ui/lint/force-warn/allowed-cli-deny-by-default-lint.rs:9]: Oh, arithmetic_overflow does not trigger with `--emit metadata`. Wtf?

Is there any deny-by-default lint that also triggers for check builds?

Comment (RalfJung) [reply src/test/ui/lint/force-warn/allowed-cli-deny-by-default-lint.rs:10]: I found a suitable lint: mutable_transmutes

File: src/test/ui/consts/assert-type-intrinsics.rs [hunk 2]
```diff
@@ -1,5 +1,3 @@
-// error-pattern: any use of this value will cause an error
```
Comment (oli-obk) [inline src/test/ui/consts/assert-type-intrinsics.rs:1]: 🎉

File: src/test/ui/consts/const-eval/const-pointer-values-in-various-types.64bit.stderr [hunk 3]
```diff
@@ -1,664 +1,258 @@
-error: any use of this value will cause an error
```
Comment (oli-obk) [inline src/test/ui/consts/const-eval/const-pointer-values-in-various-types.64bit.stderr:1]: you also need to update the 32 bit file

Comment (oli-obk) [reply src/test/ui/consts/const-eval/const-pointer-values-in-various-types.64bit.stderr:1]: ah nevermind. lol

```
// only-x86_64
// stderr-per-bitwidth
```

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
Marking for nomination for @rust-lang/lang . This PR should get a quick discussion or just an FCP as it *technically* is a breaking change (libraries can, before this PR, expose constants that will hard error on any use site).

This being a lint instead of an error has been a major pain for const eval and there is no benefit for users (in fact, diagnostics are sometimes worse if the lint is not denied).

**Comment by oli-obk** [comment]
@bors try

let's crater it quickly, but considering it has been a future incompat error before this PR, I don't expect any real breakage

**Comment by oli-obk** [comment]
@craterbot check

**Comment by RalfJung** [comment]
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-102091/retry-regressed-list.txt

**Comment by RalfJung** [comment]
- https://crater-reports.s3.amazonaws.com/pr-102091-1/try%23dbe54b90b471ee86a3b006973bf90c6cbb93415f/gh/7sDream.dotnes/log.txt fails because of a `dent(const_err)`, now the lint does not exist any more; `deny(warnings)` turns that warning into an error
- https://crater-reports.s3.amazonaws.com/pr-102091-1/try%23dbe54b90b471ee86a3b006973bf90c6cbb93415f/reg/ark-linear-sumcheck-0.3.0/log.txt similar, due to `deny(renamed_and_removed_lints`)
- https://crater-reports.s3.amazonaws.com/pr-102091-1/try%23dbe54b90b471ee86a3b006973bf90c6cbb93415f/reg/ark-poly-commit-0.3.0/log.txt similar
- https://crater-reports.s3.amazonaws.com/pr-102091-1/try%23dbe54b90b471ee86a3b006973bf90c6cbb93415f/reg/elasticsearch_exporter-0.21.0/log.txt similar
- https://crater-reports.s3.amazonaws.com/pr-102091-1/try%23dbe54b90b471ee86a3b006973bf90c6cbb93415f/reg/heapless-0.7.16/log.txt similar
- https://crater-reports.s3.amazonaws.com/pr-102091-1/try%23dbe54b90b471ee86a3b006973bf90c6cbb93415f/reg/simdutf8-0.1.4/log.txt has an `allow(const_err)` but does not actually show any const_err lints

The rest is spurious.

**Comment by joshtriplett** [comment]
We planned on this a long time ago when we started down this road, so let's see if we have consensus for the last step:

@rfcbot merge

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by RalfJung** [comment]
(that bot command doesn't seem to do anything once FCP started?)

**Comment by RalfJung** [comment]
@oli-obk FCP is over soon. :) Can I r=you once that is the case?

**Comment by oli-obk** [comment]
r=me, but you need to run tidy and remove some stray files first :)

**Pull request merged**